### PR TITLE
run.py: expect IntegrityError during Commit insert

### PR DESCRIPTION
**Pull request by jgehrcke**
See commit msg. Addresses #809.

### CODE REVIEW

File: conbench/entities/run.py [hunk 1]
```diff
@@ -209,6 +179,117 @@ def get_baseline_id(self):
...
+                    "Could not backfill default branch commits. Ignoring error "
+                    "during backfill_default_branch_commits():  %s",
+                    exc,
+                )
+
+        # If at least a commit hash and/or repository are provided, insert
+        # that. Note(JP): wouldn't it be better if we hat the constraint that a
+        # commit entity in the database has _at least_ repository specifier and
+        # commit hash? What is the value of a commit object that only refers to
+        # a repository (and not to a commit hash)?
```
Comment (jgehrcke) [inline conbench/entities/run.py]: @jonkeane @austin3dickey would love to have quick opinion, thanks!
let's put findings here: https://github.com/conbench/conbench/issues/817

File: conbench/entities/run.py [hunk 2]
```diff
@@ -209,6 +179,117 @@ def get_baseline_id(self):
...
+                # the HTTP request processing (we're right now in the middle of
+                # processing an HTTP request with new benchmark run data).
+                log.info(
+                    "Could not backfill default branch commits. Ignoring error "
+                    "during backfill_default_branch_commits():  %s",
+                    exc,
+                )
+
+        # If at least a commit hash and/or repository are provided, insert
+        # that. Note(JP): wouldn't it be better if we hat the constraint that a
```
Comment (jonkeane) [inline conbench/entities/run.py]:
```suggestion
        # that. Note(JP): wouldn't it be better if we had the constraint that a
```

File: conbench/entities/run.py (hunk 1 above)
Comment (jonkeane) [reply conbench/entities/run.py]: I'm fine with that and can't think of bad things that would happen if it didn't. The original motivation for the looseness here was to also allow us to store results on other softwares (some where we might not have a specific hash that we know we're running). We haven't done that so far, and I suspect if (when?) we do, we'll want to create a separate table for "external benchmark" that has different constraints so we can deal with that there (or we can loosen this back if we get there.

File: conbench/entities/run.py [hunk 3]
```diff
@@ -209,6 +179,117 @@ def get_baseline_id(self):
...
+    commits) from GitHub HTTP API if possible. Exceptions during this process
+    are logged and otherwise swallowed.
+
+    Return Commit.id (DB primary key) of existing Commit entity or of newly
+    created one. Expect database collision upon insert (in this case the ID for
+    the existing commit entity is returned).
+
+    Has slightly ~unpredictable run duration as of interaction with GitHub HTTP
+    API.
+    """
```
Comment (jgehrcke) [inline conbench/entities/run.py:200]: That's roughly the level of docstringing we should have in the code base. Does not need to be perfect, but between fellow developers we have to communicate the intent, in particular about error handling.

File: conbench/entities/run.py (hunk 1 above)
Comment (jgehrcke) [reply conbench/entities/run.py]: After #817 I have now removed the comment and updated the code to be clear about the fact that here, both `sha` and `repository` are non-None.

**Comment by jgehrcke** [comment]
```
        elif sha:
            # As of input schema validation this means that both, commit has
            # and repository specifier are set. Also the database schema as of
            # the time of writing this comment requires both commit hash and
            # repo specifier to be non-null. Encode this assumption here.
>           assert sha and repository
E           AssertionError
```

Oh wow. I am glad I did put this assertion in.
```
    def test_create_allow_just_repository(self, client):
        self.authenticate(client)
        data = copy.deepcopy(self.valid_payload)
        data["run_id"] = _uuid()
        data["github"]["commit"] = ""
        data["github"]["repository"] = ARROW_REPO
    
        response = client.post("/api/benchmarks/", json=data)
        new_id = response.json["id"]
        benchmark_result = BenchmarkResult.one(id=new_id)
        assert benchmark_result.run.commit.sha == ""
>       assert benchmark_result.run.commit.repository == ARROW_REPO
E       AssertionError: assert '' == 'https://github.com/apache/arrow'
E         - https://github.com/apache/arrow
```

File: conbench/entities/run.py (hunk 2 above)
Comment (jgehrcke) [reply conbench/entities/run.py]: :) :tophat:

**Pull request merged**